src/t_cose_encrypt_dec.c

```diff
@@ -0,0 +1,467 @@
+        ciphertext_len = Cipher.val.string.len;
+        detached_mode = false;
+    } else {
+        ciphertext = detached_ciphertext;
+        ciphertext_len = detached_ciphertext_len;
+        detached_mode = true;
+    }
+
+    /* Two key distribution mechanisms are supported, namely
+     *  - Direct key distribution (where no recipient info is included)
+     *  - HPKE-based key distribution (which requires recipient info)
+     */
+    if (me->key_distribution == T_COSE_KEY_DISTRIBUTION_DIRECT) {
+        if (kid_cbor.len == 0 ||
+            strncmp(me->kid.ptr, kid_cbor.ptr, me->kid.len) != 0
+           ) {
+                return(T_COSE_ERR_UNKNOWN_KEY);
+        }
+    } else {
+        /* Recipients */
+        QCBORDecode_EnterArray(&DC, NULL);
+
+        /* protected header */
+        result = QCBORDecode_GetNext(&DC, &Item);
+
+        if (result != QCBOR_SUCCESS) {
+            return(T_COSE_ERR_CBOR_FORMATTING);
+        }
+
+        if (Item.uDataType != QCBOR_TYPE_BYTE_STRING) {
+             return(T_COSE_ERR_PARAMETER_CBOR);
+        }
+
+        if (protected_hdr.uDataType != QCBOR_TYPE_BYTE_STRING) {
+             return(T_COSE_ERR_PARAMETER_CBOR);
+        }
+
+        /* Re-initialize to parse protected header */
+        QCBORDecode_Init(&DC2,
+                         (UsefulBufC)
+                         {
+                          Item.val.string.ptr,
+                          Item.val.string.len
+                         },
+                         QCBOR_DECODE_MODE_NORMAL);
+
+        QCBORDecode_EnterMap(&DC2, NULL);
+
+        /* Retrieve algorithm */
+        QCBORDecode_GetInt64InMapN(&DC2, COSE_HEADER_PARAM_ALG, &alg);
+
+        result = QCBORDecode_GetError(&DC2);
+
+        if (result != QCBOR_SUCCESS) {
+             return(T_COSE_ERR_CBOR_MANDATORY_FIELD_MISSING);
+        }
+
+        QCBORDecode_ExitMap(&DC2);
+
+        result = QCBORDecode_Finish(&DC2);
+
+        if (result != QCBOR_SUCCESS) {
+            return(T_COSE_ERR_CBOR_FORMATTING);
+        }
+
+        /* Setting key distribution parameters. */
+        switch(alg) {
+        case COSE_ALGORITHM_HPKE_P256_HKDF256_AES128_GCM:
+            key_bitlen = 128;
+            break;
+
+        case COSE_ALGORITHM_HPKE_P521_HKDF512_AES256_GCM:
+            key_bitlen = 256;
+           break;
+
+        default:
+            return(T_COSE_ERR_UNSUPPORTED_KEY_EXCHANGE_ALG);
+        }
+
+        /* unprotected header */
+        QCBORDecode_EnterMap(&DC, NULL);
+
+        /* get ephemeral */
+        QCBORDecode_GetByteStringInMapN(&DC,
+                                        COSE_HEADER_ALG_PARAM_EPHEMERAL_KEY,
+                                        &ephemeral);
+
+        result = QCBORDecode_GetError(&DC);
+
+        if (result != QCBOR_SUCCESS) {
+            return(T_COSE_ERR_CBOR_MANDATORY_FIELD_MISSING);
+        }
+
+        /* Decode ephemeral */
+        QCBORDecode_Init(&DC2,
```

Comment:
This should turn into a COSE_Key decoder function in a separate file that supports encoding and decoding COSE_Key.